### PR TITLE
Added Optimizer Nadamax

### PR DESCRIPTION
**Pull request by sourabhvarshney111**
I have added the code for NadaMax Update. I have not added the reference Paper. Let me know which one to put?

### CODE REVIEW

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    double beta1T1 = beta1 * (1 - (0.5 *
+        std::pow(0.96, (iteration + 1) * scheduleDecay)));
+
+    cumBeta1 *= beta1T;
+
+    const double biasCorrection1 = 1.0 - cumBeta1;
+
+    const double biasCorrection2 = 1.0 - (cumBeta1 * beta1T1);
+
+    if ((biasCorrection1 != 0) && (biasCorrection2 != 0))
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Can you use ``{`` around the if condition to improve the readability?

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+   * Construct the Nadam update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient
+   * @param scheduleDecay The decay parameter for decay coefficients
+   */
+  NadaMaxUpdate(const double epsilon = 1e-8,
+              const double beta1 = 0.9,
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Can you align ``beta1, beta2, scheduleDecay`` with the first parameter?

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+              const double beta2 = 0.99,
+              const double scheduleDecay = 4e-3)
+      :epsilon(epsilon),
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Pedantic style issue; can you put ``:`` after the closing ``)``?

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+namespace mlpack {
+namespace optimization {
+
+/**
+ * NadaMax is an optimizer that combines the AdaMax and NAG.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @misc{Dozat2015,
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: We can use the same paper as for the ``NadamUpdate`` class.

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+ *   address     = {},
+ *   year        = {}
+ * }
+ * @endcode
+ */
+class NadaMaxUpdate
+{
+ public:
+  /**
+   * Construct the Nadam update policy with the given parameters.
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Should be ``NadaMax``.

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp (hunk 1 above)
Comment (sourabhvarshney111) [reply src/mlpack/core/optimizers/adam/nadamax_update.hpp]: I have put the braces { after the condition. Let me know if you were telling about the condition and not what I have done.

Comment (zoq) [reply src/mlpack/core/optimizers/adam/nadamax_update.hpp]: That's what I meant, thanks. I will check the update formula one more time, and I think afterwards this is ready.

Comment (sourabhvarshney111) [reply src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Thanks

File: src/mlpack/core/optimizers/adam/nadamax_update.hpp [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    u *= beta2;
+    u = arma::max(u, arma::abs(gradient));
```
Comment (zoq) [inline src/mlpack/core/optimizers/adam/nadamax_update.hpp]: I think ``u = arma::max(u * beta2, arma::abs(gradient));`` might be easier to read, what you think?

Comment (sourabhvarshney111) [reply src/mlpack/core/optimizers/adam/nadamax_update.hpp]: Ya, that would be better for reading purpose.

**Comment by sourabhvarshney111** [comment]
No problem. New year time, of course. Happy new year

**Comment by sourabhvarshney111** [comment]
: )

**Comment by zoq** [comment]
Thanks again for another great contribution.

**Comment by rcurtin** [comment]
I made some really minor style and grammar fixes in 4c6e2dac2; let me know if there are any problems.  Thanks for the contribution! :)

**Comment by sourabhvarshney111** [comment]
Thanks for the changes. :)

**Pull request merged**